Objects/typeobject.c

```diff
@@ -426,14 +419,12 @@ assign_version_tag(struct type_cache *cache, PyTypeObject *type)
     if (!_PyType_HasFeature(type, Py_TPFLAGS_READY))
         return 0;
 
-    type->tp_version_tag = next_version_tag++;
-    /* for stress-testing: next_version_tag &= 0xFF; */
-
-    if (type->tp_version_tag == 0) {
-        // Wrap-around or just starting Python - clear the whole cache
-        type_cache_clear(cache, 1);
+    if (next_version_tag == 0) {
+        /* We have run out of version numbers */
```

Comment:
Can you please elaborate the comment to explain what happen on next_version_tag overflow? I understand that old types keep their caches, but types which are are changed are no longer cached.

Reply:
Random thought which I haven't verified:
Why not call `PyType_Modified` on base object type and clear cache on overflow, then continue incrementing version tag counter again?

That way even after overflow, new version tags are assigned. Just that every `UINT32_TMAX` type cache miss event will cause temporarily bad performance. But that seems worth it vs bad performance all the way after we hit `UINT32_TMAX`.

Reply:
> Can you please elaborate the comment to explain what happen on next_version_tag overflow? I understand that old types keep their caches, but types which are are changed are no longer cached.

It doesn't overflow. Once `next_version_tag` reaches 0, we stop issuing new versions.

Reply:
> Random thought which I haven't verified:
> Why not call `PyType_Modified` on base object type and clear cache on overflow, then continue incrementing version tag counter again?
> 
> That way even after overflow, new version tags are assigned. Just that every `UINT32_TMAX` type cache miss event will cause temporarily bad performance. But that seems worth it vs bad performance all the way after we hit `UINT32_TMAX`.

We would also need to find and clear every specialized instruction that uses the version tag. It is possible, but I don't think it is worth it. 2**32 is a large number.